skimage/util/_backends.py

```diff
@@ -0,0 +1,153 @@
+    no_dispatching = os.environ.get("SKIMAGE_NO_DISPATCHING", False)
+    if no_dispatching == "1":
+        return True
+    else:
+        return False
+
+
+def get_module_name(func):
+    """Get the public module name of a scikit-image function.
+
+    This computes the name of the public submodule in which the function can
+    be found.
+    """
+    full_name = func.__module__
+    # This relies on the fact that scikit-image does not use
+    # sub-submodules in its public API, except in one case.
+    # This means that public name can be atmost `skimage.foobar`
+    # for everything else
+    if full_name.startswith("skimage.filters.rank"):
+        public_name = "skimage.filters.rank"
+    else:
+        public_name = ".".join(full_name.split(".")[:2])
+
+    # It would be nice to sanity check things by doing something like the
+    # following. However we can't because this code is executed while the
+    # module is being imported, which means this would create a circular
+    # import
+    # mod = importlib.import_module(public_name)
+    # assert getattr(mod, func.__name__) is func
+
+    return public_name
+
+
+@cache
+def all_backends():
+    """List all installed backends and information about them"""
+    backends = {}
+    backends_ = _entry_points(group="skimage_backends")
+    backend_infos = _entry_points(group="skimage_backend_infos")
+
+    for backend in backends_:
+        backends[backend.name] = {"implementation": backend}
+        try:
+            info = backend_infos[backend.name]
+            # Double () to load and then call the backend information function
+            backends[backend.name]["info"] = info.load()()
+        except KeyError:
+            pass
+
+    return backends
+
+
+def dispatchable(func):
+    """Mark a function as dispatchable.
+
+    When a decorated function is called the installed backends are
+    searched for an implementation. If no backend implements the function
+    then the scikit-image implementation is used.
+    """
+    func_name = func.__name__
+    func_module = get_module_name(func)
+
+    # If no backends are installed at all or dispatching is disabled,
+    # return the original function. This way people who don't care about it
+    # don't see anything related to dispatching
+    if disable_dispatching() or not all_backends():
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Backends are tried in alphabetical order, this makes things
+        # predictable and stable across runs. Might need a better solution
+        # when it becomes common that users have more than one backend
+        # that would accept a call.
+        for name in sorted(all_backends()):
+            backend = all_backends()[name]
+            # If we have a BackendInformation object we check if the
+            # function we are looking for is implemented in the backend
+            if "info" in backend:
```

Comment:
The main reason I made it optional was because someone mentioned that is how it is in NetworkX.

Now that you ask, I kinda think we should make it required. It will make life simpler and it seems like a "why wouldn't you implement this" kind of thing.

Reply:
Happy to make it mandatory. :)

Reply:
the `info`(i.e. the second ep) here and in networkx are different. In NetworkX, it is just for getting the info that gets displayed on the website and the global configs for the backend. But, here I think `info` is being used to check if a function is implemented in the backend or not. In networkx, this check is done by checking for the function in the namespace that the first ep point refers to.